### PR TITLE
Improve error for incompatible generic arguments for `WeakRef`

### LINKED ISSUE #11833: Add a compiler message that `WeakRef` doesn't apply to modules
I got tripped up this morning because it looks like `WeakRef` can't apply to module references. I assume that's inherent to what modules are, but the error message provided was generic and a while to figure out.
```crystal
require "weak_ref"

module Thing
end

class A
  include Thing
end

class B
  @test : WeakRef(Thing)

  getter test

  def initialize(with_thing)
    @test = WeakRef(Thing).new(with_thing)
  end
end

v = A.new
x = B.new(v)

puts x.test.value # Error: can't cast Pointer(Void) to (Thing | Nil)
```
Once you know why the error's showing up, it makes sense, but I think it'd be helpful to include a *why* here, at least to catch similar mistakes:
```crystal
class B
  @test : WeakRef(Thing) # Error: can't create a weak reference to a module
end
```

### PR DESCRIPTION
**Pull request by caspiano**
Closes #11833

### CODE REVIEW

**Comment by Blacksmoke16** [comment]
I'd prob also add something to the docs to make this more clear before you go to do it. E.g. via the `WARNING` admonition

File: src/weak_ref.cr [hunk 1]
```diff
@@ -12,6 +14,8 @@ class WeakRef(T)
   @target : Void*
 
   def initialize(target : T)
+    {% raise "Cannot create a WeakRef to a module" if T.module? %}
```
Comment (Sija) [inline src/weak_ref.cr:17]:
```suggestion
    {% raise "Cannot create a WeakRef of a module" if T.module? %}
```

**Comment by HertzDevil** [comment]
The proper check is probably `T <= Reference?`. A similar one is used by `Box`(T).

**Review by straight-shoota — APPROVED** [review]
I think this is good enough for now. Having better error messages for other incompatible types would be nice, but we can add that later.

**Review by oprypin — APPROVED** [review]

**Comment by HertzDevil** [comment]
I believe standard library specs are only allowed to use `compile_file`, not `assert_error`.

**Pull request merged**